service/s3/s3manager/upload.go

```diff
@@ -504,20 +520,67 @@ func (a completedParts) Len() int           { return len(a) }
 func (a completedParts) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a completedParts) Less(i, j int) bool { return *a[i].PartNumber < *a[j].PartNumber }
 
+// request all the parts that are known to be completed for this upload
+func (u *multiuploader) requestCompletedParts() error {
+	var partNumberMarker *int64
+
+	for {
+		req := u.cfg.S3.ListPartsRequest(&s3.ListPartsInput{
+			Bucket:           u.uploader.in.Bucket,
+			Key:              u.uploader.in.Key,
+			UploadId:         &u.uploader.cfg.ResumeUploadID,
+			PartNumberMarker: partNumberMarker,
+		})
+
+		output, err := req.Send(u.ctx)
+		if err != nil {
+			return err
+		}
+
+		for _, part := range output.Parts {
+			completed := s3.CompletedPart{ETag: part.ETag, PartNumber: part.PartNumber}
+			u.parts = append(u.parts, completed)
+		}
+
+		if !*output.IsTruncated {
+			break
+		} else {
+			partNumberMarker = output.NextPartNumberMarker
+		}
+	}
+
+	return nil
+}
+
 // upload will perform a multipart upload using the firstBuf buffer containing
 // the first chunk of data.
 func (u *multiuploader) upload(firstBuf io.ReadSeeker) (*UploadOutput, error) {
 	params := &s3.CreateMultipartUploadInput{}
 	awsutil.Copy(params, u.in)
 
-	// Create the multipart
-	req := u.cfg.S3.CreateMultipartUploadRequest(params)
-	req.ApplyOptions(u.cfg.RequestOptions...)
-	resp, err := req.Send(u.ctx)
-	if err != nil {
-		return nil, err
+	// Fetch the completed parts when resuming an upload.
+	// The ResumeUploadID is only used if the completed parts could be fetched.
+	if len(u.uploader.cfg.ResumeUploadID) > 0 {
+		if err := u.requestCompletedParts(); err == nil {
+			u.uploadID = u.uploader.cfg.ResumeUploadID
```

Comment:
If the user provided a UploadID to resume, and that ID either doesn't exist, or an error is encountered while getting the parts, the Uploader should return an error instead of creating a new upload. This is to prevent unexpected behavior, because the upload id didn't exist already.

Reply:
ResumeUpload() will return an error now in this situation.